File: evidence/v1_5/ai_benchmark/subjects/primary/r2/pool/cpp/main.cpp
#include <cstdint>
#include <iostream>
#include <queue>
#include <string>
#include <vector>

struct Slot {
    bool occupied = false;
    uint64_t gen = 0;
    std::vector<uint8_t> data;
};

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::vector<Slot> slots;
    std::priority_queue<uint64_t, std::vector<uint64_t>, std::greater<uint64_t>> freeSlots;

    std::string out;
    out.reserve(1 << 20);

    auto isLive = [&](uint64_t i, uint64_t g) -> bool {
        return i < slots.size() && slots[i].occupied && slots[i].gen == g;
    };

    auto freeSlot = [&](uint64_t i) {
        slots[i].occupied = false;
        slots[i].data.clear();
        slots[i].data.shrink_to_fit();
        freeSlots.push(i);
    };

    std::string cmd;
    while (std::cin >> cmd) {
        if (cmd == "new") {
            uint64_t S;
            std::cin >> S;
            uint64_t i;
            if (!freeSlots.empty()) {
                i = freeSlots.top();
                freeSlots.pop();
            } else {
                i = slots.size();
                slots.emplace_back();
            }
            slots[i].occupied = true;
            slots[i].gen += 1;
            slots[i].data.assign(S, 0);
            out += "h ";
            out += std::to_string(i);
            out += ' ';
            out += std::to_string(slots[i].gen);
            out += '\n';
        } else if (cmd == "set") {
            uint64_t I, G, OFF, V;
            std::cin >> I >> G >> OFF >> V;
            if (!isLive(I, G)) {
                out += "stale\n";
            } else if (OFF >= slots[I].data.size()) {
                out += "bounds\n";
            } else {
                slots[I].data[OFF] = static_cast<uint8_t>(V);
                out += "ok\n";
            }
        } else if (cmd == "sum") {
            uint64_t I, G;
            std::cin >> I >> G;
            if (!isLive(I, G)) {
                out += "stale\n";
            } else {
                uint64_t s = 0;
                for (uint8_t b : slots[I].data) s += b;
                out += std::to_string(s);
                out += '\n';
            }
        } else if (cmd == "move") {
            uint64_t I, G, J, H;
            std::cin >> I >> G >> J >> H;
            if (!isLive(I, G) || !isLive(J, H)) {
                out += "stale\n";
            } else if (I == J && G == H) {
                out += "same\n";
            } else {
                auto &src = slots[I].data;
                auto &dst = slots[J].data;
                dst.insert(dst.end(), src.begin(), src.end());
                uint64_t newSize = dst.size();
                freeSlot(I);
                out += "ok ";
                out += std::to_string(newSize);
                out += '\n';
            }
        } else if (cmd == "free") {
            uint64_t I, G;
            std::cin >> I >> G;
            if (!isLive(I, G)) {
                out += "stale\n";
            } else {
                freeSlot(I);
                out += "ok\n";
            }
        }
    }

    uint64_t liveCount = 0;
    uint64_t liveBytes = 0;
    for (auto &s : slots) {
        if (s.occupied) {
            liveCount += 1;
            liveBytes += s.data.size();
        }
    }
    out += "live ";
    out += std::to_string(liveCount);
    out += ' ';
    out += std::to_string(liveBytes);
    out += '\n';

    std::cout << out;
    return 0;
}

File: evidence/v1_5/ai_benchmark/subjects/primary/r2/pool/rust/main.rs
use std::collections::BTreeSet;
use std::io::{self, Read, Write};

struct Slot {
    gen: u64,
    data: Option<(Vec<u8>, u64)>, // (bytes, sum of byte values)
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();

    let mut slots: Vec<Slot> = Vec::new();
    let mut free_set: BTreeSet<usize> = BTreeSet::new();

    let mut live_count: u64 = 0;
    let mut live_bytes: u64 = 0;

    let mut out = String::new();

    for line in input.lines() {
        let mut it = line.split(' ').filter(|s| !s.is_empty());
        let cmd = match it.next() {
            Some(c) => c,
            None => continue,
        };

        match cmd {
            "new" => {
                let s: usize = it.next().unwrap().parse().unwrap();
                let idx = if let Some(&i) = free_set.iter().next() {
                    free_set.remove(&i);
                    i
                } else {
                    slots.push(Slot { gen: 0, data: None });
                    slots.len() - 1
                };
                slots[idx].gen += 1;
                slots[idx].data = Some((vec![0u8; s], 0));
                live_count += 1;
                live_bytes += s as u64;
                out.push_str(&format!("h {} {}\n", idx, slots[idx].gen));
            }
            "set" => {
                let i: usize = it.next().unwrap().parse().unwrap();
                let g: u64 = it.next().unwrap().parse().unwrap();
                let off: u64 = it.next().unwrap().parse().unwrap();
                let v: u64 = it.next().unwrap().parse().unwrap();
                let live = i < slots.len() && slots[i].gen == g && slots[i].data.is_some();
                if !live {
                    out.push_str("stale\n");
                } else {
                    let (buf, sum) = slots[i].data.as_mut().unwrap();
                    let off_usize = usize::try_from(off).ok();
                    if off_usize.is_none() || off_usize.unwrap() >= buf.len() {
                        out.push_str("bounds\n");
                    } else {
                        let off_usize = off_usize.unwrap();
                        let old = buf[off_usize] as u64;
                        buf[off_usize] = v as u8;
                        *sum = *sum - old + v;
                        out.push_str("ok\n");
                    }
                }
            }
            "sum" => {
                let i: usize = it.next().unwrap().parse().unwrap();
                let g: u64 = it.next().unwrap().parse().unwrap();
                let live = i < slots.len() && slots[i].gen == g && slots[i].data.is_some();
                if !live {
                    out.push_str("stale\n");
                } else {
                    let (_, sum) = slots[i].data.as_ref().unwrap();
                    out.push_str(&format!("{}\n", sum));
                }
            }
            "move" => {
                let i: usize = it.next().unwrap().parse().unwrap();
                let g: u64 = it.next().unwrap().parse().unwrap();
                let j: usize = it.next().unwrap().parse().unwrap();
                let h: u64 = it.next().unwrap().parse().unwrap();
                let live_src = i < slots.len() && slots[i].gen == g && slots[i].data.is_some();
                let live_dst = j < slots.len() && slots[j].gen == h && slots[j].data.is_some();
                if !live_src || !live_dst {
                    out.push_str("stale\n");
                } else if i == j && g == h {
                    out.push_str("same\n");
                } else {
                    let (src_buf, src_sum) = slots[i].data.take().unwrap();
                    {
                        let (dst_buf, dst_sum) = slots[j].data.as_mut().unwrap();
                        dst_buf.extend_from_slice(&src_buf);
                        *dst_sum += src_sum;
                    }
                    free_set.insert(i);
                    live_count -= 1;
                    let new_size = slots[j].data.as_ref().unwrap().0.len();
                    out.push_str(&format!("ok {}\n", new_size));
                }
            }
            "free" => {
                let i: usize = it.next().unwrap().parse().unwrap();
                let g: u64 = it.next().unwrap().parse().unwrap();
                let live = i < slots.len() && slots[i].gen == g && slots[i].data.is_some();
                if !live {
                    out.push_str("stale\n");
                } else {
                    let (buf, _) = slots[i].data.take().unwrap();
                    live_count -= 1;
                    live_bytes -= buf.len() as u64;
                    free_set.insert(i);
                    out.push_str("ok\n");
                }
            }
            _ => {}
        }
    }

    out.push_str(&format!("live {} {}\n", live_count, live_bytes));

    io::stdout().write_all(out.as_bytes()).unwrap();
}
